### ISSUE TITLE
Workflow Headache: dev dependencies (private) are resolved when not included

### ISSUE THREAD

**Issue by ayjayt** [astral-sh/uv#10201]
Hi!

Sometimes my dev dependencies come from private repos, and therefore they can't be resolved in all workflows:

* during normal usage
* during CI testing

for example

`uv` seems to try and resolve all dependency groups even if they're not included. :-)

Just posting this here to start a discussion!

Thanks

**Comment by ayjayt** [comment]
Hi @T-256, **all dependencies get _resolved_ by default**, even if they are not actually downloaded and installed. This is not documented, but it is the actual implemented behavior.

You can test this by creating a nonsense group with a non-sense dependency and then explicitly telling it not to be installed.

**Comment by charliermarsh** [comment]
Yeah, it's intentional that we always resolve all groups. If you have workflows where you need to use your project without resolving those groups, you should be able to `uv run` without re-resolving, as long as you have a `uv.lock` checked in.

**Comment by ayjayt** [comment]
I feel I don't understand the implementation of `uv run` and `uv.lock` well enough to comment, by my understanding is that:

if there is an updated `uv.lock` checked in to the **git repo** , `uv run` will skip the resolving phase and use whatever is in `uv.lock`, even if the dependencies impossible to resolve were added after the `uv.lock` was checked in?

Is there some kind of contribution I could make to provide a solution by which certain dependency groups could be entirely ignored by `uv`?

Such as a 

```
[tool.uv]
ignore-groups = ["nickelback", "phish"]
```

It would also roll into my other issue: https://github.com/astral-sh/uv/issues/10200#issuecomment-2564096152

?

**Comment by schloerke** [comment]
> it's intentional that we always resolve all groups

Maybe I'm missing something, but when calling `uv` command where I've explicitly asked for an extra group (`--group foo`) or to ignore a group (`--no-group foo`) then I'd expect it to only attempt to resolve those tangible groups. Only the groups required for execution should be used for dependency resolution.

My interpretation of the groups is that if I stated `uv sync --no-group foo` with the pyproject.toml above, I'd expect it to have the same behavior as `[dependency-group]` project file which has called `uv sync`.  

Crafting a `uv.lock` file manually doesn't seem like a long term solution.

**Comment by zanieb** [comment]
We resolve all groups and use a shared resolution for the project (stored in `uv.lock`). Selecting groups only applies to which subset of the lockfile is synced. This is a fundamental behavior of uv — we're not likely to change it. I understand why it'd be nice to avoid resolving those groups in some cases, but it's not realistic for us to implement.

**Comment by ayjayt** [comment]
Yeah I took a look at your source code and realized it would be difficult, so I switched to commenting them out.

I think there could be such a thing as "explicit groups" that would only ever been included if they were explicitly asked to be included. I think that would play nicely with the "fork" concepts that was recently implemented in the conflicts PR.

If you view your program as a pipeline from 

specification (pyproject.toml)  --> lock file (uv.lock) --> environment (.venv)

I think that model could incorporate this idea, and the "explicit groups" would be ignored between any edge.

That said, I realize you have a lot to deal with, as do I, I wish I had time to make contribution, and I appreciate the work you've done on the project.
